projects/dataset_recorder/data_recorder.cc

```diff
@@ -0,0 +1,246 @@
+        cam->readFrame(img);
+
+    }
+
+
+}
+
+void saveThread(cv::Mat img, std::string fileName) {
+    cv::imwrite(fileName, img);
+}
+
+int bobMain(int argc, char* argv[])
+{
+
+    using degree_t = units::angle::degree_t;
+    std::vector<cv::Mat> images;
+    std::vector<std::string> fileNames;
+
+    // setting up gps
+    const char *path_linux = "/dev/ttyACM1"; // path for linux systems
+    BoBRobotics::GPS::Gps gps;
+    gps.connect(path_linux);
+    BoBRobotics::BN055 imu;
+    const int maxTrials = 3;
+    int numTrials = maxTrials;
+    std::atomic<bool> shouldQuit{false};
+    std::mutex mex;
+
+
+    time_t rawtime;
+    struct tm * timeinfo;
+    char buffer[80];
+
+    time (&rawtime);
+    timeinfo = localtime(&rawtime);
+
+    strftime(buffer,sizeof(buffer),"%d-%m-%Y %H:%M:%S",timeinfo);
+    std::string str(buffer);
+
+    int seconds_to_run = 30; // default value
+    if (argc > 1) {
+        seconds_to_run = stoi(std::string(argv[1]));
+        std::cout << "running for " << seconds_to_run << " seconds" << std::endl;
+
+    }
+
+
+
+    // check to see if we get valid coordinates by checking GPS quality
+    while (numTrials > 0) {
+        try {
+            if (gps.getGPSData().gpsQuality != BoBRobotics::GPS::GPSQuality::INVALID) {
+                std::cout << " we have a valid measurement" << std::endl;
+                break;
+            }
+        } catch(...) {
+            std::cout << " measuring failed, trying again in 1 second " << "[" << maxTrials - numTrials << "/" << maxTrials << "]" << std::endl;
+            std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+            numTrials--;
+        }
+    }
+    if (numTrials == 0) {
+        std::cout << " There is no valid gps measurement, please try waiting for the survey in to finish and restart the program " << std::endl;
+        exit(1);
+    }
+
+    auto gps_time =  gps.getGPSData().time;
+    int gps_hour = gps_time.hour;
+    int gps_minute = gps_time.minute;
+    int gps_second = gps_time.second;
+
+    std::string folderName;
+    //std::cin >> folderName;
+
+    folderName = str;
+
+    const char *c  = folderName.c_str();
+    mkdir(c, 0777);
+    std::cout << "directory " << c << " created" << std::endl;
+
+    auto cam = getPanoramicCamera();
+    const auto cameraRes = cam->getOutputSize();
+    std::cout << "camera initialised" << std::endl;
+
+    // unwrapper for camera
+   // auto unwrapper = cam->createUnwrapper(cv::Size(WIDTH,HEIGHT));
+
+
+    cv::Mat imgOrig;
+    //--------------------------------->>>>>>>>>>> START RECORDING <<<<<<<<<<<<----------------------------
+
+
+    // start camera thread
+    cv::Mat tmpImage, unwrappedImage, resized;
+    cv::Mat outputImage(cv::Size(WIDTH,HEIGHT), CV_8UC3);
+    std::thread readCameraThread(&readCameraThreadFunc, cam.get(), std::ref(tmpImage), std::ref(shouldQuit), std::ref(mex));
+    std::this_thread::sleep_for(std::chrono::milliseconds(3000));
```

Comment:
Presumably this sleep is to wait for the camera to come alive? I think it would be much better to have the camera in the main thread (as I assume it's faster) and have a worker thread for the GPS

Reply:
Yes, the sleep is for waiting for the camera.
I was thinking about the camera in the main thread and the gps as the worker thread, but I could not figure out how to do it. 
With the current method,  it is easy to just get the current image when the main loop get to the line.
Doing the other way around, I would have to start a stopwatch in the main thread for timing the gps and also to write to the csv which I found a bit clunky. What do you think?